### ISSUE TITLE
Implementing or porting an accurate, SIMD optimized IDCT algorithm

### ISSUE THREAD

**Issue by antonfirsov** [SixLabors/ImageSharp#306]
We are constantly (re)evaluating the design and implementation plans for our Jpeg decoder, but one thing is sure: we can't go without a SIMD-optimized and accurate IDCT (Inverse [Discrete Cosine Transform](https://en.wikipedia.org/wiki/Discrete_cosine_transform)) implementation.

*The issue description is in WIP state. I'm trying to add notes here while doing research.*

### What we currently have:
- [A floating point method](https://github.com/SixLabors/ImageSharp/blob/jpeg-lab/src/ImageSharp/Formats/Jpeg/GolangPort/Components/DCT.cs#L52) ported from an [SSE C++ implementation](https://github.com/norishigefukushima/dct_simd/blob/master/dct/dct8x8_simd.cpp#L239).
  - This one is very fast but seems to be inaccurate
- [A standard + unoptimized integer method](https://github.com/SixLabors/ImageSharp/blob/jpeg-lab/src/ImageSharp/Formats/Jpeg/GolangPort/Components/DCT.cs#L52) ported from libjpeg.
  - This one seems to be accurate and standard-comforming

### What we need
A standard conformant, accurate implementation, that's fast enough. Most likely a floating point one.
Good candidates:
- A SIMD-optimized variant of the [standard floating-point method](https://github.com/mozilla/mozjpeg/blob/master/jidctflt.c)
  - Based on: *Arai, Agui, and Nakajima: "A fast DCT-SQ scheme for images"*
  - Ported to assembly in [libjpeg-turbo](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/simd/jidctflt-sse2.asm)
- The default [integer](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/simd/jidctint-sse2.asm) [methods](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/simd/jidctint-sse2-64.asm) used by libjpeg-turbo

We need to SIMD-optimize our color conversion code as well, and we should use `Vector<float>` or `Vector4` for that. I think, we should expect better accuracy and performance, if our entire post processing chain is done with floats, so I would prefer floating point IDCT implementations over integer ones.

### On .NET SIMD API-s
- I don't know when do the [new .NET SIMD](https://github.com/dotnet/corefx/issues/15957) API-s land. Even if ETA is not that much, I'm afraid using them will hurt the portability and cross-platform nature of ImageSharp. So, unless someone has a good reason, I prefer not using them.
- In my experience, even `Vector<T>` is not mature enough at the moment. SIMD support for different methods is not well documented. It's easy to loose it, eg. by using `Vector<T>` as a struct member. Some reverse-engineering hints on this topic [could be found here](https://github.com/SixLabors/ImageSharp/pull/274#issuecomment-314309643).
- In short: I think having an implementation that uses nothing more than the plain-old `Vector4` should be good enough, but correct me if I am wrong! :)

### Testing conformance and accuracy
We should unit-test our algorithms against reference implentations in the following way:
![image](https://user-images.githubusercontent.com/6835152/29642904-af836fd0-886b-11e7-8a28-cf97f5310864.png)

We have accurate, mathematically correct [FDCT/IDCT reference implementations](https://github.com/SixLabors/ImageSharp/blob/jpeg-lab/tests/ImageSharp.Tests/Formats/Jpg/Utils/ReferenceImplementations.AccurateDCT.cs) in our test code base. We should always test the new ones against them.

**Comment by KLuuKer** [comment]
+1 on the actual\trusted implementation it would allow for easy testing if a new\existing piece of code is correct.

Maybe if there is a good way to detect what features\types are available we could switch between different algorithms automatically (and trough configuration of course).
That is if you want to maintain all those different pieces of code.

**Comment by antonfirsov** [comment]
@KLuuKer this modularity is actually what I plan. Our users don't need to know about the backing IDCT algorithms. We can introduce something easy for them though, like:
```C#
public enum PerformanceMode
{
    Fast,
    Accurate
}
```

**Comment by antonfirsov** [comment]
I'm closing this because there is nothing wrong with the accuracy of our IDCT implementation. All I needed to do was [this](https://github.com/SixLabors/ImageSharp/blob/1448e6bac1be6309f704a149ec71dbf32171e360/src/ImageSharp/Formats/Jpeg/GolangPort/Components/Decoder/JpegBlockPostProcessor.cs#L81).

Inaccurate is the new accurate!

**Comment by tocsoft** [comment]
too accurate, got to laugh. We should probably add a flag to toggle compatiblity vs accuracy.
